File: src/java/magmac/api/iter/Iters.java
package magmac.api.iter;

import magmac.api.Option;
import magmac.api.head.EmptyHead;
import magmac.api.head.HeadedIter;
import magmac.api.head.RangeHead;
import magmac.api.head.SingleHead;

/**
 * Factory methods for creating {@link Iter} instances.
 */

public final class Iters {
    private static <T> Iter<T> fromArray(T[] array) {
        return new HeadedIter<>(new RangeHead(array.length))
                .map((Integer index) -> array[index]);
    }

    /**
     * Creates an iterator from an {@link Option}, yielding zero or one element.
     */
    public static <T> Iter<T> fromOption(Option<T> option) {
        return option
                .<Iter<T>>map((T t) -> new HeadedIter<>(new SingleHead<>(t)))
                .orElseGet(() -> new HeadedIter<>(new EmptyHead<>()));
    }

    /**
     * Creates an iterator from the provided values.
     */
    public static <T> Iter<T> fromValues(T... values) {
        return Iters.fromArray(values);
    }

    /**
     * Returns an iterator with no elements.
     */
    public static <T> Iter<T> empty() {
        return new HeadedIter<>(new EmptyHead<>());
    }
}

File: src/web/magmac/api/iter/Iters.ts
import { Option } from "../../../magmac/api/Option";
import { EmptyHead } from "../../../magmac/api/head/EmptyHead";
import { HeadedIter } from "../../../magmac/api/head/HeadedIter";
import { RangeHead } from "../../../magmac/api/head/RangeHead";
import { SingleHead } from "../../../magmac/api/head/SingleHead";
export class Iters {
	private static fromArray( array : T[]) : Iter<T> {return new HeadedIter<>( new RangeHead( array.length)).map( 0);;}
	public static fromOption( option : Option<T>) : Iter<T> {return option.<Iter<T>>map( 0).orElseGet( 0);;}
	public static fromValues( ...values : T[]) : Iter<T> {return Iters.fromArray( values);;}
	public static empty() : Iter<T> {return new HeadedIter<>( new EmptyHead<>( ));;}
}
